### PR TITLE
Added a field identifier to the User Collection 

### PR DESCRIPTION
**Pull request by shankeleven**
**What kind of change does this PR introduce?**

feature

Issue Number: #1554 

**Did you add tests for your changes?**

yes

**Summary**

added a new field named identifier to the **User**, which is
- unique
- numeric
- immutable
- incremental & sequential
- automatically generated for every user when registering
- the count for identifier persists between server restarts

**Does this PR introduce a breaking change?**

no

### CODE REVIEW

**Comment by shankeleven** [comment]
@Cioppolo14 ma'am please review this pull request

**Comment by Cioppolo14** [comment]
@shankeleven Can you fix the failing linting related tests?

**Comment by shankeleven** [comment]
> @shankeleven Can you fix the failing linting related tests?

sure ma'am

**Comment by shankeleven** [comment]
ma'am regarding these linting errors , only one of these 8 has been written by me for this issue
and it is regarding treating the error as any 
which the eslint doesn't support but is in my opinion good for handling errors of all sort
it would have been a problem if it was any other variable
feel free to advise if i have some wrong idea about this

**Comment by shankeleven** [comment]
@palisadoes @xoldyckk @Cioppolo14  please assign some reviewers to this PR

**Comment by Cioppolo14** [comment]
@shankeleven Sorry I missed your response.  Palisadoes is not the main reviewer of the pull requests. We have a set of volunteer reviewers, but we try to do a check first to make sure we are as efficient with everyone's time as possible.  Feel free to @ me, or even find me on slack if something gets lost.  Please fix the failing & skipped tests, then we will assign reviewers.

File: src/models/User.ts
Comment (SiddheshKukade) [inline src/models/User.ts:287]: Why have you removed required from this ?

File: tests/resolvers/User/user.spec.ts
Comment (SiddheshKukade) [inline tests/resolvers/User/user.spec.ts:30]: Remove commented code is not useful for future

**Review by SiddheshKukade — CHANGES_REQUESTED** [review]

File: tests/resolvers/User/user.spec.ts
Comment (shankeleven) [reply tests/resolvers/User/user.spec.ts]: i have fixed both sir
if there are any further changes required , I'd be happy to do them.

File: src/models/User.ts
Comment (shankeleven) [reply src/models/User.ts]: fixed it sir

File: src/models/User.ts
Comment (xoldyckk) [inline src/models/User.ts:150]: make this field required, not null

File: src/models/User.ts
Comment (xoldyckk) [inline src/models/User.ts:245]: remove this field it's unnecessary

File: src/models/identifier_count.ts
Comment (xoldyckk) [inline src/models/identifier_count.ts:26]: don't understand what's being done here, explain please

File: src/models/User.ts
Comment (xoldyckk) [inline src/models/User.ts:287]: unnecessary

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by shankeleven** [comment]
@xoldyckk please let me know if there are any more changes required

File: src/models/identifier_count.ts
Comment (SiddheshKukade) [reply src/models/identifier_count.ts:26]: This looks like some wrapper function to me @xoldyckk

**Review by SiddheshKukade — APPROVED** [review]
This PR looks fine to merge and it also satisfies the requirement of a ID. 
Screenshot 2024-02-18 at 10 49 12 AM

**Comment by SiddheshKukade** [comment]
@shankeleven please take a pull from origin develop

**Comment by SiddheshKukade** [comment]
@shankeleven please fix the linting errors.

**Comment by shankeleven** [comment]
> @shankeleven please take a pull from origin develop

done sir



> @shankeleven please fix the linting errors.

these linting errors are from files that i have not edited and are not relevant to this issue
so i thought it wouldn't be good to change anything in the files not related to the issue
i could do it if permitted though.

**Comment by palisadoes** [comment]
We reverted the PR cussing the linting problem. Update with your upstream

**Comment by palisadoes** [comment]
If merging with the latest upstream doesn't work, then please lint the offending file.

**Comment by palisadoes** [comment]
Please fix the linting errors so we can merge this

**Comment by shankeleven** [comment]
> Please fix the linting errors so we can merge this

sir,  i tried correcting it but couldn't do it
so i guess i will need some guidance on how to navigate through it
as even if i copy paste the file (facing linting errors ) from the main repo
and ofcourse the git pull coudn't fix this either
these linting errors don't seem to go away

**Comment by palisadoes** [comment]
If you click on the link of the failed tests, you'll see this. You'll need to edit these files to fix the linting. Sometimes they are files that you didn't edit directly, but ones that were affected by merge conflicts. We cannot merge code that hasn't been fully tested

image

**Comment by xoldyckk** [comment]
@shankeleven seems mongodb doesn't support auto-incrementing ids at the database level, so you're using this workaround to implement the same logic in talawa-api

why can't you simply fetch the user with greatest `identifier` and assign the user being created an identifier value of `fetched user identfier + 1`? why a seperate schema for `identifier` is needed here, care to explain your rationale behind this?

**Comment by shankeleven** [comment]
> seems mongodb doesn't support auto-incrementing ids at the database level, so you're using this workaround to implement the same logic in talawa-api
> 
> @shankeleven why can't you simply fetch the user with greatest `identifier` and assign the user being created an identifier value of `fetched user identfier + 1`? why a seperate schema for `identifier` is needed here, care to explain your rationale behind this?

Sure sir, there are actually a few reasons behind chosing this specific approach, some fundamental and some for long term. 

- when multiple requests are being processed simultaneously, fetching the user with the greatest identifier and then incrementing it can lead to race conditions. In a high-concurrency environment, multiple requests might fetch the same greatest identifier and attempt to use it, leading to conflicts and as a result the identifier may end up not being unique which would cause errors.

- Fetching the user with the greatest identifier every time a new user is created can be less efficient, especially as the database grows and the number of users increases. 

- As the platform scales and we distribute our database across multiple servers or shards, managing a separate schema for the identifier count provides better scalability options. It ensures that the identifier generation process remains consistent and reliable across distributed environments.

**Comment by xoldyckk** [comment]
@shankeleven there are two seperate mutation operations to the database, since it's not a transaction atomicity isn't guaranteed, does mongodb support transactions? is it doable?

File: src/models/User.ts
Comment (xoldyckk) [inline src/models/User.ts:300]: can you link me some docs about this? i'm not familiar with this

File: src/models/identifier_count.ts
Comment (xoldyckk) [inline src/models/identifier_count.ts:1]: wouldn't `counters` be a better name for this collection?
for reference:- https://www.mongodb.com/docs/v2.2/tutorial/create-an-auto-incrementing-field/#auto-increment-counters-collection
they're making use of `userId` to target counter that is used for user collection

**Comment by palisadoes** [comment]
@shankeleven Please address the questions from @xoldyckk

**Comment by palisadoes** [comment]
@shankeleven 

1. We'd like to merge this PR. 
2. Please go through the review questions, update the PR as requested and mark them as resolved when completed

**Comment by palisadoes** [comment]
Please also fix the failing tests (linting)

**Comment by shankeleven** [comment]
> @shankeleven there are two seperate mutation operations to the database, since it's not a transaction atomicity isn't guaranteed, does mongodb support transactions? is it doable?

sorry for overlooking the question
yes mongodb does support transactions
but my main rationale behind not implementing the transactions was to reduce the resources overload that comes along the transactions
as even in the worst case if a registration fails for the user and the identifier still increases, we would still be fulfilling all the essential criteria for the field *identifier* **i.e.** incremental , numeric , unique
and besides, we would anyways have the count of total registered users available all the time
so i thought there was no need to implement a transaction here.

**Comment by DMills27** [comment]
@shankeleven Can you fix the failing tests?

**Comment by palisadoes** [comment]
This is an update on the PR merging freeze:

1. In the next few hours we will be merging the `develop-userTypeFix` branch into the `develop` branch. 
1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. This merge will cause some conflicts in your PR.

We decided to do this at the beginning of the weekend to give us all time to adjust PR code and create bug fixes that may arise.

Update your code at or after midnight GMT on the morning of March 23, 2024. (5:30am IST). 

If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by Cioppolo14** [comment]
@shankeleven Please fix the conflicting files and failed tests.

**Comment by palisadoes** [comment]
Closing as abandoned

**Comment by shankeleven** [comment]
quite saddening to see it closed as it took a decent amount of thought for proper implementation
but fair enough.

**Comment by palisadoes** [comment]
We were going through the PRs, evaluating those that were probably abandoned. I'll reopen it.

**Comment by Cioppolo14** [comment]
@shankeleven Please addressing the conflicting files and open comments.

**Comment by Cioppolo14** [comment]
@shankeleven Are you still working on this PR?

**Comment by palisadoes** [comment]
Closing as abandoned

**Comment by shankeleven** [comment]
apologies for not being able to complete

**Pull request closed without merging**
